src/main/java/net/fabricmc/loom/configuration/CompileConfiguration.java

```diff
@@ -235,20 +240,91 @@ private Path getLockFile() {
 		return cacheDirectory.resolve("." + pathHash + ".lock");
 	}
 
-	private boolean getAndLock() {
-		final Path lock = getLockFile();
+	enum LockResult {
+		// acquired immediately or after waiting for another process to release
+		ACQUIRED_CLEAN,
+		// already owned by current pid
+		ACQUIRED_ALREADY_OWNED,
+		// acquired due to current owner not existing
+		ACQUIRED_PREVIOUS_OWNER_MISSING
+	}
 
-		if (Files.exists(lock)) {
-			return true;
-		}
+	private LockResult acquireProcessLockWaiting(Path lockFile) {
+		// one hour
+		return this.acquireProcessLockWaiting(lockFile, 1000L * 60 * 60);
+	}
 
+	private LockResult acquireProcessLockWaiting(Path lockFile, long timeoutMs) {
 		try {
-			Files.createFile(lock);
-		} catch (IOException e) {
-			throw new UncheckedIOException("Failed to acquire getProject() configuration lock", e);
+			return this.acquireProcessLockWaiting_(lockFile, timeoutMs);
+		} catch (final IOException e) {
+			throw new RuntimeException("Exception acquiring lock " + lockFile, e);
+		}
+	}
+
+	// Returns true if our process already owns the lock
+	@SuppressWarnings("BusyWait")
+	private LockResult acquireProcessLockWaiting_(Path lockFile, long timeoutMs) throws IOException {
+		final Logger logger = Logging.getLogger("loom_acquireProcessLockWaiting");
+		final long currentPid = ProcessHandle.current().pid();
+		boolean abrupt = false;
+
+		if (Files.exists(lockFile)) {
+			long lockingProcessId;
+
+			try {
+				lockingProcessId = Long.parseLong(Files.readString(lockFile));
+			} catch (final Exception e) {
+				lockingProcessId = -1;
+			}
+
+			if (lockingProcessId == currentPid) {
+				return LockResult.ACQUIRED_ALREADY_OWNED;
+			}
+
+			logger.lifecycle("Lock file '{}' is currently held by pid '{}'.", lockFile, lockingProcessId);
```

Comment:
As commented bellow im a little hesitant to point the user towards the lock file, I worry if they start tampering with it they may end up in more of a mess. Maybe just print the project name that goes into the lock file?